packages/wrangler/e2e/wrangler-mixed-mode-resources.test.ts

```diff
@@ -0,0 +1,309 @@
+				remote: true,
+			},
+		}),
+		expectedResponseMatch: "This is a response from Workers AI",
+		// AI bindings work without mixed mode flag
+		worksWithoutMixedMode: true,
+	},
+	{
+		name: "KV",
+		scriptPath: "kv.js",
+		setup: async (helper) => {
+			const ns = await helper.kv(false);
+			await helper.run(
+				`wrangler kv key put --remote --namespace-id=${ns} test-mixed-mode-key existing-value`
+			);
+			return { id: ns };
+		},
+		generateWranglerConfig: ({ id: namespaceId }) => ({
+			name: "mixed-mode-kv-test",
+			main: "kv.js",
+			compatibility_date: "2025-01-01",
+			kv_namespaces: [
+				{
+					binding: "KV_BINDING",
+					id: namespaceId,
+					remote: true,
+				},
+			],
+		}),
+		expectedResponseMatch: "The pre-existing value is: existing-value",
+	},
+	{
+		name: "R2",
+		scriptPath: "r2.js",
+		setup: async (helper) => {
+			await helper.seed({ "test.txt": "existing-value" });
+			const name = await helper.r2(false);
+			await helper.run(
+				`wrangler r2 object put --remote ${name}/test-mixed-mode-key --file test.txt`
+			);
+			onTestFinished(async () => {
+				await helper.run(
+					`wrangler r2 object delete --remote ${name}/test-mixed-mode-key`
+				);
+			});
+			return { name };
+		},
+		generateWranglerConfig: ({ name: bucketName }) => ({
+			name: "mixed-mode-r2-test",
+			main: "r2.js",
+			compatibility_date: "2025-01-01",
+			r2_buckets: [
+				{
+					binding: "R2_BINDING",
+					bucket_name: bucketName,
+					remote: true,
+				},
+			],
+		}),
+		expectedResponseMatch: "The pre-existing value is: existing-value",
+	},
+	{
+		name: "D1",
+		scriptPath: "d1.js",
+		setup: async (helper) => {
+			await helper.seed({
+				"schema.sql": dedent`
+          CREATE TABLE entries (key TEXT PRIMARY KEY, value TEXT);
+          INSERT INTO entries (key, value) VALUES ('test-mixed-mode-key', 'existing-value');
+        `,
+			});
+			const { id, name } = await helper.d1(false);
+			await helper.run(
+				`wrangler d1 execute --remote ${name} --file schema.sql`
+			);
+			return { id, name };
+		},
+		generateWranglerConfig: ({ id, name }) => ({
+			name: "mixed-mode-d1-test",
+			main: "d1.js",
+			compatibility_date: "2025-01-01",
+			d1_databases: [
+				{
+					binding: "DB",
+					database_id: id,
+					database_name: name,
+					remote: true,
+				},
+			],
+		}),
+		expectedResponseMatch: "existing-value",
+	},
+];
+
+describe("Wrangler Mixed Mode E2E Tests", () => {
+	describe.each(testCases)("$name", (testCase) => {
+		let helper: WranglerE2ETestHelper;
+
+		beforeEach(() => {
+			helper = new WranglerE2ETestHelper();
+		});
+
+		it("works with mixed mode enabled", async () => {
+			await helper.seed(
+				path.resolve(__dirname, "./seed-files/mixed-mode-workers")
+			);
+
+			const setupResult = (await testCase.setup?.(helper)) ?? {};
+
+			const wranglerConfig = testCase.generateWranglerConfig(setupResult);
+			await helper.seed({
+				"wrangler.json": JSON.stringify(wranglerConfig, null, 2),
+			});
```

Comment:
Every test does these three statements. Can we either combine them into one property in TestCase or have a helper that does this?